### PR TITLE
Create user login, signup, logout, forget pass, update profile features

### PR DESCRIPTION
**Pull request by arisachue**
- Log in feature uses Firebase authentication
- User can log in with email and pass
- User can sign up with email and pass
- Can click forgot password to have a password reset link sent to their email
- Can also update their email and password
- Users can log out

Sign up page:
Screen Shot 2022-07-12 at 2 34 50 PM

Profile page:
Screen Shot 2022-07-12 at 2 35 15 PM

Update profile page:
Screen Shot 2022-07-12 at 2 35 53 PM

Log in page:
Screen Shot 2022-07-12 at 2 34 20 PM

Forgot password page:
Screen Shot 2022-07-12 at 2 36 26 PM
Screen Shot 2022-07-12 at 2 37 25 PM

### CODE REVIEW

File: src/components/App/App.jsx [hunk 1]
```diff
@@ -53,21 +58,21 @@ const Operations = Object.freeze({
 })
 
 // get best recipe matches with user's food items
```
Comment (reddysindhu1) [inline src/components/App/App.jsx:60]: I think you might have accidentally left this effect commented out in this commit

File: src/contexts/AuthContext.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { auth } from "../firebase"
+
+const AuthContext = createContext()
+
+export function AuthProvider({ children }) {
+    const [currentUser, setCurrentUser] = useState({})
+    const [loading, setLoading] = useState(true)
+
+    function signup(email, password) {
+        console.log("here")
```
Comment (reddysindhu1) [inline src/contexts/AuthContext.jsx]: Remove the leftover console.log

File: src/contexts/AuthContext.jsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    useEffect(() => {
+        const unsubscribe = onAuthStateChanged(auth, (user) => {
+            setCurrentUser(user)
+            setLoading(false)
+        })
+
+        return () => {
+            unsubscribe()
+        }
+    }, [])
```
Comment (reddysindhu1) [inline src/contexts/AuthContext.jsx:44]: I'm curious, do you not need to add auth or user into this array? Do you not want to rerun the effect when one of those values change?

File: src/components/ForgotPass/ForgotPass.jsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        }
+        setLoading(false)
+    }
+
+    return (
+        <div className="home">
+            <Card>
+            <Card.Body>
+                <h2>Reset Password</h2>
+                {message && <Alert variant="success">{message}</Alert>}
```
Comment (reddysindhu1) [inline src/components/ForgotPass/ForgotPass.jsx:34]: Do you set message to null? If not, the alert will always be there right?

File: src/components/App/App.jsx [hunk 5]
```diff
@@ -53,21 +58,21 @@ const Operations = Object.freeze({
...
+//       const recipeApiIngredients = listRecipeUrl + ingredientParams
+//       var { data } = await axios(recipeApiIngredients)
+//       setRecipes(data)
+//     } catch (err) {
+//       //TODO: error handling
+//       console.log("error in fetching from backend")
+//     }
+//   }
+//   fetchData()
+// }, [products])
```
Comment (mjquigley) [inline src/components/App/App.jsx:75]: Should this be uncommented?

File: src/components/Home/Home.jsx [hunk 6]
```diff
@@ -1,11 +0,0 @@
...
-import { Link } from "react-router-dom"
-import "./Home.css"
-
-export default function Home() {
-  return (
-    <nav className="home">
-        <p>Home</p>
-    </nav>
-  )
-}
```
Comment (mjquigley) [inline src/components/Home/Home.jsx:11]: nit: this change isn't related to user login, it could be in it's own PR

File: src/components/PrivateRoute/PrivateRoute.jsx [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import React from "react"
+import { Route, Navigate } from "react-router-dom"
+import { useAuth } from "../../contexts/AuthContext"
+
+const PrivateRoute = ({ children }) => {
```
Comment (mjquigley) [inline src/components/PrivateRoute/PrivateRoute.jsx:5]: Should any of the existing routes be PrivateRoutes?

File: src/components/LogIn/LogIn.jsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    const [loading, setLoading] = useState(false)
+    const navigate = useNavigate()
+
+    const handleSubmit = async (e) => {
+        e.preventDefault()
+
+        try {
+            setLoading(true)
+            await login(emailRef.current.value, passwordRef.current.value)
+            navigate("/profile")
```
Comment (mjquigley) [inline src/components/LogIn/LogIn.jsx:21]: Should it redirect to the profile page or the page they were previously on?

File: src/components/UpdateProfile/UpdateProfile.jsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
+import * as React from "react"
```
Comment (mjquigley) [inline src/components/UpdateProfile/UpdateProfile.jsx:1]: nit: the profile page could probably have been a separate PR.

File: src/firebase.jsx [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { initializeApp } from "firebase/app"
+// import "firebase/auth"
```
Comment (mjquigley) [inline src/firebase.jsx]: nit: delete

**Review by mjquigley — APPROVED** [review]

**Comment by mjquigley** [comment]
Looks great! Firebase seems really nice

File: src/contexts/AuthContext.jsx (hunk 3 above)
Comment (arisachue) [reply src/contexts/AuthContext.jsx:44]: `onAuthStateChanged()` is an event listener so it will continue to check if the auth state changes. I only have to run that once in the beginning, but the auth changes as different users log in

File: src/components/ForgotPass/ForgotPass.jsx (hunk 4 above)
Comment (arisachue) [reply src/components/ForgotPass/ForgotPass.jsx:34]: When Alert is an empty string, nothing shows up. I think bootstrap does that check for me

File: src/components/LogIn/LogIn.jsx (hunk 8 above)
Comment (arisachue) [reply src/components/LogIn/LogIn.jsx:21]: I'm planning to have the user logged in before using the website, so once they sign in, they would always see the profile page (and can now click on the other pages)
